Introduce quota-aware filesystem ES plugin

pugnascotia opened the pull request:
Closes #61145.

This PR adds a quota-aware filesystem plugin to Elasticsearch. This plugin offers a way to provide user quota limits (specifically, total quota size and available quota size) to Elasticsearch, in an implementation-agnostic manner.

As part of this work, this PR also introduces the concept of "bootstrap only" plugins, which are excluded from the normal plugin loading process.

Finally, note that this implementation supports `createLink(...)`, since ES / Lucene use hard links where possible.

pugnascotia committed:
First import

pugnascotia committed:
WIP - moving quota-aware-fs to a plugin

pugnascotia committed:
Check test now passes

pugnascotia committed:
Introduce bootstrap-only plugins

pugnascotia committed:
WIP - trying to write an IT

pugnascotia committed:
Fruitless tweaks

pugnascotia committed:
Add logic to enable the plugin

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into quota-aware-fs

pugnascotia committed:
Fixes to get ES running with quota enforcement

pugnascotia committed:
Always require system property to set quota fs properties URI

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into quota-aware-fs

pugnascotia committed:
WIP - packaging test for quota fs plugin

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into quota-aware-fs

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into quota-aware-fs

pugnascotia committed:
Cleanups

pugnascotia committed:
Remove unnecessary test file

pugnascotia committed:
Also test refreshes

pugnascotia committed:
Tweak

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into quota-aware-fs

pugnascotia committed:
Add comment

pugnascotia committed:
Revert changes to run pkg tests on Darwin

pugnascotia committed:
More cleanup

pugnascotia committed:
Add test

pugnascotia committed:
Support creating symlinks in quota-aware fs plugin

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Plugins)

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesExtension.java:
```diff
@@ -105,6 +107,14 @@ public void setHasNativeController(boolean hasNativeController) {
         this.hasNativeController = hasNativeController;
     }
 
+    public boolean isBootstrapOnly() {
```
This naming is a bit confusing, specifically the "only" part. Other plugins are not "bootstrap" plugins, so this plugin is not "only" a bootstrap plugin. Perhaps instead we could have a "type" attribute or something like that which defaults to "isolated", and then have this alternative value "bootstrap"?

pugnascotia replied:
I've changed to a "type" as you suggested.

rjernst commented on server/src/main/java/org/elasticsearch/plugins/BootstrapPlugin.java:
```diff
@@ -0,0 +1,27 @@
+ * Bootstrap plugins are not loaded during normal Elasticsearch operation. This
+ * interface exists to make this explicit.
+ */
+public interface BootstrapPlugin {
```
I don't understand what advantages a marker interface gives here. A bootstrap plugin's purpose is to add code to the root classloader, so that it can be used by the jdk and other parts of the system that are loaded before plugins are loaded. I don't see a reason why these plugins should have a Plugin class at all?

pugnascotia replied:
Ah...I think I originally added it, thinking I'd use it to filter out bootstrap plugins, but of course you need to load the plugin to inspect it 🤦. I should have removed it earlier, but it's gone now.

rjernst commented on distribution/src/bin/elasticsearch:
```diff
@@ -54,6 +54,14 @@ fi
 
+# If the quote-aware filesystem plugin is installed, then we need to pass extra
+# flags to Java in order to use the plugin.
+if [[ -d "$ES_HOME/plugins/quota-aware-fs" ]]; then
```
The idea of bootstrap plugins is to avoid this type of specialized logic (https://github.com/elastic/elasticsearch/issues/29306). Instead, the plugin should define which system properties need to be added, and bin/elasticsearch should introspect those upon finding all bootstrap plugins.

pugnascotia replied:
I've added infrastructure to do this, see what you think.

What I haven't done, however, is update `bin/elasticsearch.bat` (yet). My batch knowledge it pretty woeful, so if we want to support this mechanism on Windows, I'll need to pair with someone who knows what they're doing 😅

rjernst reviewed (commented):
Thanks for putting this together @pugnascotia. I especially appreciate the packaging tests.

With that said, while I do want to see bootstrap plugins implemented, I thought the quota aware fs was going to be a one off until they could be properly implemented. I've left a few comments as to why I think the bootstrap plugins here are not what was originally envisioned.

pugnascotia committed:
Generalise the bootstrap plugin mechanism

pugnascotia committed:
Use enums for plugin type

pugnascotia committed:
Prohibit isolated plugins from specifying java.opts at runtime

pugnascotia committed:
Compilation fix

pugnascotia committed:
Checkstyle

pugnascotia committed:
Formatting

rjernst commented on server/src/main/java/org/elasticsearch/plugins/PluginInfo.java:
```diff
@@ -172,36 +183,45 @@ public static PluginInfo readFromProperties(final Path path) throws IOException
+
+        final String typeString = propsMap.remove("type");
+        if (Strings.isNullOrEmpty(typeString)) {
+            throw new IllegalArgumentException("property [type] is missing for plugin [" + name + "]");
```
I think this will break a lot of plugin authors who have static plugin properties. Can we instead default type to isolated?

pugnascotia replied:
It now defaults to isolated.

rjernst commented on server/src/main/java/org/elasticsearch/plugins/PluginInfo.java:
```diff
@@ -172,36 +183,45 @@ public static PluginInfo readFromProperties(final Path path) throws IOException
 
+        final PluginType type = PluginType.valueOf(typeString.toUpperCase());
+
+        final String javaOpts = propsMap.remove("java.opts");
```
Can we error if java opts is set for non bootstrap plugins?

pugnascotia replied:
I actually had that in, then removed it because of the validation I added to `bin/elasticsearch`, but I've put it back in to be on the safe side.

rjernst commented on server/src/main/java/org/elasticsearch/plugins/PluginsService.java:
```diff
@@ -349,31 +350,36 @@ static void checkForFailedPluginRemovals(final Path pluginsDirectory) throws IOE
-            bundles.add(bundle);
+            final Bundle bundle = readPluginBundle(plugin, type);
+            if (bundle.plugin.getType() == PluginType.BOOTSTRAP) {
+                logger.trace("--- skipping bootstrap only [{}] [{}]", type, plugin.toAbsolutePath());
```
nit: only -> plugin

rjernst commented on server/src/main/java/org/elasticsearch/plugins/PluginsService.java:
```diff
@@ -349,31 +350,36 @@ static void checkForFailedPluginRemovals(final Path pluginsDirectory) throws IOE
         }
 
+        logger.trace(
+            "findBundles(" + type + ") returning: " + bundles.stream().map(b -> b.plugin.getName()).sorted().collect(Collectors.toList())
```
nit: this trace is going to produce a string every time. the message should be wrapped in a lambda?

pugnascotia replied:
I've wrapped it in a lambda.

rjernst reviewed (commented):
Looking better. A few more comments. Also, for the bat file, perhaps @williamrandolph can help there.

pugnascotia committed:
Lots of compilation fixes, apply review feedback

pugnascotia committed:
Add tests

pugnascotia commented on buildSrc/src/main/resources/plugin-descriptor.properties:
```diff
@@ -43,3 +43,12 @@ extended.plugins=${extendedPlugins}
+#
+# 'java.opts': any additional command line parameters to pass to the JVM when
+# Elasticsearch starts. Only applies when 'type' is 'bootstrap'.
+<%= type == "bootstrap" ? "" : "# " %>java.opts=${javaOpts}
```
It is. Perhaps it's overkill, but I thought it was worth some extra discouragement from anyone setting this field. I can remove it if you feel it's too much.

rjernst replied:
I'm not understand why it would be commented. If it is a bootstrap plugin, the java.opts should be added. If it is an isolated plugin, then why add java.opts at all, commented or otherwise?

pugnascotia commented on server/src/main/java/org/elasticsearch/plugins/PluginInfo.java:
```diff
@@ -172,36 +183,45 @@ public static PluginInfo readFromProperties(final Path path) throws IOException
+                              classname, extendedPlugins, hasNativeController, type, javaOpts);
+    }
+
+    private static boolean parseBooleanValue(String name, String rawValue) {
```
`Boolean.parseBoolean` never throws an exception, so we'd lose the validation part.

rjernst replied:
I didn't mean the parse method from `Boolean`, I mean that from `o.e.c.Booleans`, which throws an error on non true/false values.

pugnascotia commented:
@rjernst As an alternative to updating the `.bat` file, should we consider moving the `java` CLI options logic to a Java class, e.g. `JvmOptionsParser` or similar?

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into quota-aware-fs

pugnascotia committed:
Don't load quota-aware-fs plugin for docs:integTest

pugnascotia commented:
@elasticmachine update branch

rjernst commented:
> As an alternative to updating the .bat file, should we consider moving the java CLI options logic to a Java class

That might work. The more we move to java, the easier it is to coalesce them when eventually moving to a launcher JVM.

pugnascotia committed:
Address review feedback

pugnascotia committed:
Process bootstrap plugin in JvmOptionsParser

pugnascotia committed:
Refactor BootstrapJvmOptions for unit testing

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into quota-aware-fs

pugnascotia committed:
Checkstyle

pugnascotia committed:
Use java.nio.file due to checkstyle

pugnascotia committed:
Windows fix

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into quota-aware-fs

pugnascotia committed:
Fix test

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into quota-aware-fs

pugnascotia committed:
Also update elasticsearch-service.bat

beiske commented on plugins/quota-aware-fs/src/main/java/org/elasticsearch/fs/quotaaware/QuotaAwareFileSystemProvider.java:
```diff
@@ -0,0 +1,464 @@
+            }
+
+            try {
+                // Inlined org.apache.lucene.util.IOUtils.close(...) to
```
Now that this is becoming a plugin and as such is released for every ES version, it could just use the Lucene version in ES

pugnascotia replied:
Ah, I meant to do that, thanks Konrad.

pugnascotia committed:
Fix serialisation

pugnascotia committed:
De-inline Lucene's IOUtils.close

pugnascotia committed:
Test fix

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into quota-aware-fs

pugnascotia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into quota-aware-fs

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into quota-aware-fs

pugnascotia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into quota-aware-fs

pugnascotia committed:
Ignore quota-aware-fs in qa:smoke-test-plugins

pugnascotia committed:
Yet more test fixes

pugnascotia commented:
@elasticmachine run elasticsearch-ci/1
@elasticmachine run elasticsearch-ci/2

pugnascotia commented:
@elasticmachine run elasticsearch-ci/1

pugnascotia committed:
And another test fix

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into quota-aware-fs

pugnascotia commented:
@rjernst CI is now green.

pugnascotia commented:
BTW I realised I haven't written docs yet - I'll sort that out.

pugnascotia committed:
Add docs for quota-aware-fs plugin

elasticmachine commented:
Pinging @elastic/es-docs (>docs)

pugnascotia commented:
I added docs under a new "filesystem plugins" section - none of the existing sections seemed like a good fit.

jrodewig commented on docs/plugins/quota-aware-fs.asciidoc:
```diff
@@ -0,0 +1,42 @@
+total and available amounts of disk space.
+
+To use the plugin, install it on all nodes and restart them. You must configure the plugin
+by supplying the `es.fs.quota.file` <<jvm-options,JVM system property>> on startup. This
```
```suggestion
by supplying the `es.fs.quota.file` {ref}/jvm-options.html[JVM system property] on startup. This
```

pugnascotia committed:
Update docs/plugins/quota-aware-fs.asciidoc

Fix ref syntax

Co-authored-by: James Rodewig <40268737+jrodewig@users.noreply.github.com>

pugnascotia committed:
Prevent fs activity causing flaky quota tests

The test QuotaAwareFileSystemProviderTests.testFileStoreNotLimited
relied on the host's disk space totals not changing in the course of the
test. This has mostly always been true, but then failed in CI. Prevent
this from reoccurring by introducing a FilesystemProvider that takes
a snapshot of the disk totals upon creation.

pugnascotia committed:
Checkstyle

pugnascotia commented:
@elasticmachine run elasticsearch-ci/1
@elasticmachine run elasticsearch-ci/bwc

rjernst commented on plugins/quota-aware-fs/src/main/java/org/elasticsearch/fs/quotaaware/QuotaAwareFsPlugin.java:
```diff
@@ -0,0 +1,32 @@
+ * used space can perform accurate calculation with respect to the disk
+ * space that is actually available to Elasticsearch.
+ */
+public class QuotaAwareFsPlugin extends Plugin {
```
This class shouldn't be necessary anymore?

pugnascotia replied:
Good spot, I've removed it.

rjernst commented on plugins/quota-aware-fs/build.gradle:
```diff
@@ -0,0 +1,6 @@
+esplugin {
+  description 'A bootstrap plugin that adds support for interfacing with filesystem that enforce user quotas.'
+  classname 'org.elasticsearch.fs.quotaaware.QuoteAwareFsPlugin'
```
This shouldn't be necessary for bootstrap plugins, we should change the validation to only require it for isolated plugins.

pugnascotia replied:
I've changed the validation and removed this line. I initially retained an empty `classname` in the properties file as it felt more consistent, but then realised that it would be more consistent to omit it since we omit `javaOpts` for bootstrap plugins. I'm partly being thrown off by all the "example plugin" preamble at the top of the file. Do we really want that there? We could use JSP comments to remove it from the build output?

rjernst replied:
At one point the template file was also referenced from our docs as an example for plugin authors who built the plugin properties on their own. I don't have any preference on how it is done, but we should definitely continue to have such an example.

rjernst commented on qa/smoke-test-plugins/build.gradle:
```diff
@@ -33,6 +33,10 @@ testClusters.integTest {
       return
     }
+    if (pluginName == 'quota-aware-fs') {
+      // This plugin has to be configured to work via system properties
```
We should just be skipping all bootstrap plugins rather than hardcoding `quota-aware-fs`?

rjernst commented on server/src/main/java/org/elasticsearch/plugins/PluginInfo.java:
```diff
@@ -49,6 +50,8 @@
     public static final String ES_PLUGIN_PROPERTIES = "plugin-descriptor.properties";
     public static final String ES_PLUGIN_POLICY = "plugin-security.policy";
 
+    private static final Version QUOTA_FS_PLUGIN_SUPPORT = Version.CURRENT;
```
This should be 7.11?

pugnascotia replied:
I can't do that until the changes are backported - I learnt that the hard way on another branch.

rjernst replied:
This can be done. There are 2 ways to handle the backports. One is to put CURRENT in master, change it on the backport, then go back to master and change all the occurrences. The other is to put the intended version in master and disable bwc tests (there is a global flag in the root build.gradle), backport, then re-enable bwc tests in master after the backport is complete. The latter has the advantages that the backport can be tested locally before committing to master, as well as only a single place to change (the bwc flag) on followup, instead of potentially multiple versioned conditional statements.

pugnascotia replied:
I believe this is the only place that I reference a version. I'll go for the former route I think.

rjernst commented on x-pack/qa/smoke-test-plugins/build.gradle:
```diff
@@ -21,6 +21,10 @@ testClusters.integTest {
       // Do not attempt to install ingest-attachment in FIPS 140 as it is not supported (it depends on non-FIPS BouncyCastle)
       return
     }
+    if (pluginName == 'quota-aware-fs') {
```
Same comment as before. Anywhere we are iterating over plugins like this it should be for isolated plugins? All bootstrap plugins will require packaging tests.

pugnascotia replied:
I talked with Rene about this, but it turned out not to be trivial. Maybe we can revisit this if/when we add another bootstrap plugin?

rjernst reviewed (commented):
A few nits

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into quota-aware-fs

pugnascotia committed:
Remove QuotaAwareFsPlugin, drop classname req for bootstrap plugins

pugnascotia commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginType.java:
```diff
@@ -0,0 +1,34 @@
+
+import java.util.Locale;
+
+public enum PluginType {
```
Docs added.

pugnascotia committed:
Omit classname from bootstrap plugin descriptor

pugnascotia committed:
Handle missing plugin classname when loading plugins
